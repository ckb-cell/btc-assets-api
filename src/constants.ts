export const CUSTOM_HEADERS = {
  ApiCache: 'x-api-cache',
  ResponseCacheable: 'x-response-cacheable',
};

export enum ApiCacheStatus {
  Hit = 'HIT',
  Miss = 'MISS',
}

export const JWT_IGNORE_URLS = ['/token', '/docs'];
