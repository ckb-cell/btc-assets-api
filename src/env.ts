import 'dotenv/config';
import z from 'zod';
import process from 'node:process';
import { omit } from 'lodash';

const envSchema = z.object({
  NODE_ENV: z.string().default('development'),
  PORT: z.string().optional(),
  ADDRESS: z.string().optional(),
  NETWORK: z.enum(['mainnet', 'testnet']).default('testnet'),

  DOMAIN: z.string().optional(),

  SENTRY_DSN_URL: z.string().optional(),
  REDIS_URL: z.string(),
  RATE_LIMIT_PER_MINUTE: z.number().default(100),
  LOGGER_LEVEL: z.enum(['debug', 'info', 'warn', 'error']).default('info'),

  ADMIN_USERNAME: z.string().optional(),
  ADMIN_PASSWORD: z.string().optional(),

  /**
   * JWT_SECRET is used to sign the JWT token for authentication.
   */
  JWT_SECRET: z.string(),
  /**
   * The URL/USERNAME/PASSWORD of the Bitcoin JSON-RPC server.
   * The JSON-RPC server is used to query the Bitcoin blockchain.
   */
  BITCOIN_JSON_RPC_URL: z.string(),
  BITCOIN_JSON_RPC_USERNAME: z.string(),
  BITCOIN_JSON_RPC_PASSWORD: z.string(),
  /**
   * The URL of the Electrs API.
   * Electrs is a Rust implementation of Electrum Server.
   * It is used to query the Bitcoin blockchain (balance, transactions, etc).
   */
  BITCOIN_ELECTRS_API_URL: z.string(),

  /**
   * Bitcoin SPV service URL
   * https://github.com/ckb-cell/ckb-bitcoin-spv-service
   */
  BITCOIN_SPV_SERVICE_URL: z.string(),
  /**
   * The URL of the CKB JSON-RPC server.
   */
  CKB_RPC_URL: z.string(),
  /**
   * Paymaster private key, used to sign the transaction with paymaster cell.
   */
  PAYMASTER_PRIVATE_KEY: z.string(),
  /**
   * Paymaster cell capacity in shannons
  * (254 CKB for RGB++ capacity + 61 CKB for change cell capacity + 1 CKB for fee cell)
   */
  PAYMASTER_CELL_CAPACITY: z.coerce.number().default(316 * 10 ** 8),
  PAYMASTER_CELL_PRESET_COUNT: z.coerce.number().default(500),
  PAYMASTER_CELL_REFILL_THRESHOLD: z.coerce.number().default(0.3),

  UNLOCKER_CRON_SCHEDULE: z.string().default('*/5 * * * *'),
  UNLOCKER_CELL_BATCH_SIZE: z.coerce.number().default(100),
  UNLOCKER_MONITOR_SLUG: z.string().default('btctimelock-cells-unlock'),

  TRANSACTION_QUEUE_JOB_DELAY: z.coerce.number().default(120 * 1000),
});

export type Env = z.infer<typeof envSchema>;
export const env = envSchema.parse(process.env);

export const getSafeEnvs = () =>
  omit(env, ['ADMIN_PASSWORD', 'JWT_SECRET', 'BITCOIN_JSON_RPC_PASSWORD', 'PAYMASTER_PRIVATE_KEY']);

export const isGenerateTokenPrivate = env.NODE_ENV === 'production' && env.ADMIN_USERNAME && env.ADMIN_PASSWORD;
