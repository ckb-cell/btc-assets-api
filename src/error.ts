export enum AppErrorCode {
  UnknownResponseError = 0x9999,
}
