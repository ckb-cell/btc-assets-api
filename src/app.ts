import { FastifyInstance } from 'fastify';
import { AxiosError } from 'axios';
import bitcoinRoutes from './routes/bitcoin';
import { env } from './env';
import tokenRoutes from './routes/token';
import swaggerPlugin from './plugins/swagger';
import jwtPlugin from './plugins/jwt';
import cachePlugin from './plugins/cache';
import rateLimitPlugin from './plugins/rate-limit';
import * as Sentry from '@sentry/node';

if (env.SENTRY_DSN_URL && env.NODE_ENV !== 'development') {
  Sentry.init({
    dsn: env.SENTRY_DSN_URL,
    tracesSampleRate: 1.0,
  });
}

async function routes(fastify: FastifyInstance) {
  fastify.register(rateLimitPlugin);
  fastify.register(swaggerPlugin);
  fastify.register(jwtPlugin);
  fastify.register(cachePlugin);

  fastify.register(tokenRoutes, { prefix: '/token' });
  fastify.register(bitcoinRoutes, { prefix: '/bitcoin/v1' });

  fastify.setErrorHandler((error, _, reply) => {
    Sentry.captureException(error);
    if (error instanceof AxiosError) {
      reply
        .status(error.response?.status || 500)
        .send({ ok: false, error: error.response?.data.error ?? error.message });
      return;
    }
    reply
      .status(error.statusCode ?? 500)
      .send({ ok: false, statusCode: error.statusCode, message: error.message });
  });
}

export default routes;
