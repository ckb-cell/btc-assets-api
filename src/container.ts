import { createContainer, InjectionMode, asValue, asClass, asFunction } from 'awilix';
import { Redis } from 'ioredis';
import pino from 'pino';
import Bitcoind from './services/bitcoind';
import ElectrsAPI from './services/electrs';
import { env } from './env';
import TransactionManager from './services/transaction';
import Paymaster from './services/paymaster';
import { RPC as CkbRPC, Indexer as CkbIndexer } from '@ckb-lumos/lumos';
import Unlocker from './services/unlocker';

export interface Cradle {
  env: typeof env;
  logger: pino.BaseLogger;
  redis: Redis;
  ckbRpc: CkbRPC;
  ckbIndexer: CkbIndexer;
  bitcoind: Bitcoind;
  electrs: ElectrsAPI;
  paymaster: Paymaster;
  unlocker: Unlocker;
  transactionManager: TransactionManager;
}

const container = createContainer<Cradle>({
  injectionMode: InjectionMode.PROXY,
  strict: true,
});

container.register({
  env: asValue(env),
  logger: asValue(pino()),
  redis: asValue(
    new Redis(env.REDIS_URL, {
      maxRetriesPerRequest: null,
    }),
  ),
  ckbRpc: asFunction(() => new CkbRPC(env.CKB_RPC_URL)).singleton(),
  ckbIndexer: asFunction(() => new CkbIndexer(env.CKB_INDEXER_URL)).singleton(),
  bitcoind: asClass(Bitcoind).singleton(),
  electrs: asClass(ElectrsAPI).singleton(),
  paymaster: asClass(Paymaster).singleton(),
  transactionManager: asClass(TransactionManager).singleton(),
  unlocker: asClass(Unlocker).singleton(),
});

export default container;
