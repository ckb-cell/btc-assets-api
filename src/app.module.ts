import { Module } from '@nestjs/common';
import { AppService } from './app.service';
import { TokenModule } from './modules/token/token.module';
import { BitcoinModule } from './modules/bitcoin/bitcoin.module';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { envSchema } from './env';
import { SentryModule } from '@ntegral/nestjs-sentry';
import { nodeProfilingIntegration } from '@sentry/profiling-node';
import { CacheModule } from '@nestjs/cache-manager';
import type { RedisClientOptions } from 'redis';
import * as redisStore from 'cache-manager-redis-store';
import { StoreConfig } from 'cache-manager';

@Module({
  imports: [
    ConfigModule.forRoot({
      isGlobal: true,
      validate: envSchema.parse,
    }),
    CacheModule.registerAsync<RedisClientOptions>({
      isGlobal: true,
      imports: [ConfigModule],
      inject: [ConfigService],
      useFactory: async (configService: ConfigService) =>
        ({
          store: redisStore,
          url: configService.get('REDIS_URL'),
          ttl: 10,
        }) as StoreConfig,
    }),
    SentryModule.forRootAsync({
      imports: [ConfigModule],
      inject: [ConfigService],
      useFactory: async (configService: ConfigService) => ({
        dsn: configService.get('SENTRY_DSN'),
        environment: configService.get('NODE_ENV'),
        tracesSampleRate: 0.5,
        profilesSampleRate: 0.5,
        integrations: [nodeProfilingIntegration()],
        logLevels:
          configService.get('NODE_ENV') === 'production'
            ? ['warn', 'error']
            : ['debug', 'log', 'warn', 'error'],
        beforeSend: (event) => {
          if (event.level === 'error') {
            return event;
          }
          return null;
        },
      }),
    }),
    TokenModule,
    BitcoinModule,
  ],
  controllers: [],
  providers: [AppService],
})
export class AppModule {}
