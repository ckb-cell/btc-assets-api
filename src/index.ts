import process from 'node:process';
import { env } from './env';
import { buildFastify } from './app';

const port = parseInt(env.PORT || '3000', 10);
const host = env.ADDRESS || '0.0.0.0';

const app = buildFastify();

app.listen({ port, host }, (err, address) => {
  if (err) {
    console.error(err);
    process.exit(1);
  }

  // eslint-disable-next-line no-console
  console.log(`Server listening at ${address}`);
});
