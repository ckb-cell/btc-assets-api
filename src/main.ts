import { NestFactory } from '@nestjs/core';
import { AppModule } from './app.module';
import { FastifyAdapter, NestFastifyApplication } from '@nestjs/platform-fastify';
import { DocumentBuilder, SwaggerModule } from '@nestjs/swagger';
import * as pkg from '../package.json';
import { patchNestjsSwagger } from '@anatine/zod-nestjs';

async function bootstrap() {
  const app = await NestFactory.create<NestFastifyApplication>(AppModule, new FastifyAdapter());

  const config = new DocumentBuilder()
    .setTitle('Bitcoin/RGB++ Assets API')
    .setVersion(pkg.version)
    .build();
  patchNestjsSwagger();
  const document = SwaggerModule.createDocument(app, config);
  SwaggerModule.setup('docs', app, document, {
    swaggerOptions: {
      defaultModelRendering: 'model',
      defaultModelExpandDepth: 4,
      defaultModelsExpandDepth: -1,
    },
  });

  await app.listen(3000);
}
bootstrap();
