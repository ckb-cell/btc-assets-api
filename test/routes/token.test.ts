import { expect, test } from 'vitest';
import { buildFastify } from '../../src/app';

test('`/token/generate` - successfuly', async () => {
  const fastify = buildFastify();
  await fastify.ready();

  const response = await fastify.inject({
    method: 'POST',
    url: '/token/generate',
    payload: {
      app: 'test',
      domain: 'test.com',
    },
  });
  const data = response.json();

  expect(response.statusCode).toBe(200);
  expect(data.token).toBeDefined();

  await fastify.close();
});

test('`/token/generate` - without params', async () => {
  const fastify = buildFastify();
  await fastify.ready();

  const response = await fastify.inject({
    method: 'POST',
    url: '/token/generate',
  });
  const data = response.json();

  expect(response.statusCode).toBe(400);
  expect(data.message).toMatchSnapshot();

  await fastify.close();
});

test('`/token/generate` - invalid domain', async () => {
  const fastify = buildFastify();
  await fastify.ready();

  const response = await fastify.inject({
    method: 'POST',
    url: '/token/generate',
    payload: {
      app: 'test',
      domain: '\\',
    },
  });
  const data = response.json();

  expect(response.statusCode).toBe(500);
  expect(data.message).toEqual('Failed to generate token: Invalid URL');

  await fastify.close();
});

test('`/token/generate` - with pathname', async () => {
  const fastify = buildFastify();
  await fastify.ready();

  const response = await fastify.inject({
    method: 'POST',
    url: '/token/generate',
    payload: {
      app: 'test',
      domain: 'http://test.com/abc',
    },
  });
  const data = response.json();

  expect(response.statusCode).toBe(500);
  expect(data.message).toEqual('Failed to generate token: Must be a valid domain without path');

  await fastify.close();
});

test('`/token/generate` - with protocol', async () => {
  const fastify = buildFastify();
  await fastify.ready();

  const response = await fastify.inject({
    method: 'POST',
    url: '/token/generate',
    payload: {
      app: 'test',
      domain: 'https://test.com',
    },
  });
  const data = response.json();

  expect(response.statusCode).toBe(200);
  expect(data.token).toBeDefined();

  await fastify.close();
});

test('`/token/generate` - with port', async () => {
  const fastify = buildFastify();
  await fastify.ready();

  const response = await fastify.inject({
    method: 'POST',
    url: '/token/generate',
    payload: {
      app: 'test',
      domain: 'test.com:3000',
    },
  });
  const data = response.json();

  expect(response.statusCode).toBe(200);
  expect(data.token).toBeDefined();

  await fastify.close();
});
