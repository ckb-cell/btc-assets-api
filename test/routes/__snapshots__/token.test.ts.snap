// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`\`/token/generate\` - without params 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "null",
    "path": [],
    "message": "Expected object, received null"
  }
]"
`;
